Game logic and doors generation

Ane1y opened the pull request.

Ane1y committed:
hey now we have level logic!

Ane1y committed:
Merge remote-tracking branch 'origin/implementation1-anely' into implementation1-anely

Ane1y committed:
hey now we have level logic!

Ane1y committed:
added function to check whether path from one point to another is possible

Ane1y committed:
doors are already here!!

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/GameLogic.java:
```diff
@@ -1,18 +1,44 @@
+        // map generation
+        var levelBuilder = new LevelBuilder();
+        levelBuilder.complexity(complexity);
+        state.levelMap = levelBuilder.build();
```
Можно
```java
var levelBuilder = new LevelBuilder();
state.levelMap = levelBuilder
    .complexity(complexity)
    .build();
```

Ane1y replied:
фига

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/GameLogic.java:
```diff
@@ -1,18 +1,44 @@
+        // unit generation
+        state.units.add(state.player);
+
+        var unitFactory = new UnitFactory(complexity);
+        var numberOfEnemies = (int)Math.log(state.player.getLevel());
+        for (int i = 0; i < numberOfEnemies; i++) {
```
Определение числа врагов стоит вынести куда-то в отдельную функцию.
Ещё `state.player.getLevel()` уже равно `complexity`, можно было использовать его

Ane1y committed:
added small test

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/state/Map.java:
```diff
@@ -2,24 +2,39 @@
+    public Map(int w, int h) {
 
+        this.height = h + 2;
+        this.width = w + 2;
+        this.map = new Map.MapTile[height][width];
+
```
Перепутаны ширина и высота

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/game/LevelBuilder.java:
```diff
@@ -65,6 +67,9 @@ public Map build() {
             map.setTile(0, height / 2, Map.MapTile.DOOR_IN);
         }
 
+        map.setTile(generateDoor(map), Map.MapTile.DOOR_OUT_NORMAL);
+        map.setTile(generateDoor(map), Map.MapTile.DOOR_OUT_HARD);
```
Всегда генерируются одинаковые двери

BurnBirdX7 replied:
Всё ещё одинаковые двери, но пока забьём

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/GameLogic.java:
```diff
@@ -1,18 +1,42 @@
+        int complexity = state.player.getLevel();
+
+        // map generation
+        var levelBuilder = new LevelBuilder();
+        levelBuilder.complexity(complexity);
+        state.levelMap = levelBuilder.build();
+        delta.setMap(state.levelMap);
```
Генерация карты никак не зависит от того в какую дверь мы вошли

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/game/LevelBuilder.java:
```diff
@@ -79,4 +84,17 @@ private Map buildFromDisk() {
+
+    // generates door and checks if the player can reach ot from entrance
+    private Position generateDoor(Map map) {
+        Position doorPos = getRandomPosition();
+        while(map.checkDistance(doorPos, map.getEntrance()) == -1) {
+            doorPos = getRandomPosition();
+        }
+        return doorPos;
+    }
```
Обойти комнату DFS от входа - составить список клеток досягаемых из двери. Можно отслеживать расстояние при обходе, и добавлять в список только клеты на расстоянии больше чем минимальное.

Выбирать позицию для выхода как рандомно из списка.

Ane1y replied:
бфсом!!

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/state/Map.java:
```diff
@@ -57,10 +76,12 @@ public boolean isFree(Position pos) {
-    private final MapTile[][] map;
-
-    private Position entrance = new Position();
-
+    public boolean isDoorOut(Position pos) {
+        var curTileType = getTile(pos);
```
maybe `isExit()`?

BurnBirdX7 replied:
**maybe** `isExit()`

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/state/Map.java:
```diff
@@ -74,4 +95,45 @@ public int hashCode() {
     }
+
+
+
+    private record queueNode(Position pos, int dist) {}
+
```
`QueueNode` - UpperCamelCase

BurnBirdX7 replied:
Что такое `int dist`?

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/state/Map.java:
```diff
@@ -74,4 +95,45 @@ public int hashCode() {
+
+    private record queueNode(Position pos, int dist) {}
+
+    private boolean isCellValid(int row, int col) {
+        return ((row >= 0) && (row < height) && (col >= 0) && (col < width));
+    }
```
`isCellInbound()`?

BurnBirdX7 replied:
Должно быть в терминах X, Y не col/row

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/state/Map.java:
```diff
@@ -74,4 +95,45 @@ public int hashCode() {
+        return ((row >= 0) && (row < height) && (col >= 0) && (col < width));
+    }
+
+    // returns -1 if path was not found, distance otherwise
+    public int checkDistance(Position from, Position to) {
+        // These arrays show the 4 possible movement from a cell
```
`getDistance()`?

Ane1y replied:
ладно...

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/state/Map.java:
```diff
@@ -74,4 +95,45 @@ public int hashCode() {
+        // These arrays show the 4 possible movement from a cell
+        int[] rowNum = {-1, 0, 0, 1};
+        int[] colNum = {0, -1, 1, 0};
+
+        boolean[][] visited = new boolean[height][width];
+
```
Перепутаны высота и ширина

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/state/Map.java:
```diff
@@ -74,4 +95,45 @@ public int hashCode() {
+            for (int i = 0; i < 4; i++) {
+                int row = point.getX() + rowNum[i];
+                int col = point.getY() + colNum[i];
+
+                if (isCellValid(row, col) && map[row][col] == MapTile.FLOOR && !visited[row][col]) {
```
Используй `getTiles(x, y)` вместо `map[row][col]`

BurnBirdX7 replied:
`map[row][col] == MapTile.FLOOR` также уже должно быть в функции `isFree`

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/state/Map.java:
```diff
@@ -74,4 +95,45 @@ public int hashCode() {
+    // returns -1 if path was not found, distance otherwise
+    public int checkDistance(Position from, Position to) {
+        // These arrays show the 4 possible movement from a cell
+        int[] rowNum = {-1, 0, 0, 1};
+        int[] colNum = {0, -1, 1, 0};
+
```
Работай в терминах X, Y а не столбцов/колонок

BurnBirdX7 replied:
Стоит хранить как 4 `Position`а,  а не два массива безотносительных значений (мб стоит завести класс `Movement`/`PositionShift` специально для хранения относительных позиций.

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/state/Map.java:
```diff
@@ -74,4 +95,45 @@ public int hashCode() {
+
+        while (!q.isEmpty()) {
+            queueNode u = q.poll();
+            Position point = u.pos;
+            if (point.getX() == to.getX() && point.getY() == to.getY())
+                return u.dist;
```
Сделай у `Position` метод `equals` и используй его для сравнения двух точек

BurnBirdX7 reviewed (changes requested):
Мало тестов -3 балла

BurnBirdX7 committed:
Merge branch 'implementation1' into implementation1-anely

Ane1y commented:
> Мало тестов -3 балла

у меня тестов больше чем у всех в этом проекте...

Ane1y committed:
fixed all the comments from Artemiy Lazarev
added differebt level complexity which depends on door out
now coordinates in getDistance is coordinates not rows and cols

Ane1y committed:
Merge remote-tracking branch 'origin/implementation1-anely' into implementation1-anely

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/GameLogic.java:
```diff
@@ -1,18 +1,51 @@
-        return null;
+        var delta = new Delta();
+        int complexity = state.player.getLevel();
+        switch (state.doorOut) {
+            case DOOR_OUT_HARD -> complexity = (int)Math.floor(complexity * 1.5);
+            case DOOR_OUT_TREASURE_ROOM -> complexity = 0;
+        }
```
может
```java
complexity = switch (state.doorOut) {
    case DOOR_OUT_HARD -> (int)Math.floor(complexity * 1.5);
    case DOOR_OUT_NORMAL -> complexity;
    case DOOR_OUT_TREASURE_ROOM -> 0;
}
```

BurnBirdX7 replied:
Зачем нужно поле `doorOut`, которое может хранить в себе любой тайл, а по умолчанию это вообще `DOOR_IN` (???).

Вместо этого можно использовать `state.map.getTile(state.player.getPosition())`,
если это длинно, в `State` можно добавить метод `getPlayerTile()`, который будет делать то же самое. (вместо нового поля с непонятной периодичностью обновления)

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/GameLogic.java:
```diff
@@ -1,18 +1,51 @@
+        levelBuilder.build();
+        delta.setMap(state.levelMap);
+
+        // unit generation
+        state.units.add(state.player);
+
+        var unitFactory = new UnitFactory(complexity);
```
Предыдущий список `state.units` не был очищен

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/GameLogic.java:
```diff
@@ -1,18 +1,51 @@
+
+        if (complexity == 0) {
+            //todo:generate treasures??
+        }
```
Этим уже должен был заняться `UnitFactory`

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/GameLogic.java:
```diff
@@ -1,18 +1,51 @@
+
+    private int numberOfUnit(int complexity) {
+        return (int)Math.log(complexity);
+    }
```
Должен быть зашит в `Map` как поле, значение задано `LevelBuilder`-ом

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/LevelLogic.java:
```diff
@@ -15,8 +14,9 @@ public LevelLogic(GameLogic gameLogic, State state) {
-        if (state.levelMap.getTile(state.player.getPosition()) == Map.MapTile.DOOR_OUT) {
+        var curPos = state.player.getPosition();
+        if (state.levelMap.isDoorOut(curPos)) {
+            state.doorOut = state.levelMap.getTile(curPos);
```
Тут тоже `doorOut`, этого обновления можно избежать, если завести метод вместо поля

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/game/LevelBuilder.java:
```diff
@@ -65,18 +71,31 @@ public Map build() {
 
+        map.setTile(generateDoor(map), Map.MapTile.DOOR_OUT_NORMAL);
+        map.setTile(generateDoor(map), Map.MapTile.DOOR_OUT_HARD);
+
+        numberOfEnemies = (int)Math.log(complexity);
+
```
Это должно быть поле класса `Map`, а не `LevelBuilder`. Оно может быть в `LevelBuilder`, но оно тут не очень полезно.

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/game/LevelBuilder.java:
```diff
@@ -65,18 +71,31 @@ public Map build() {
     }
+    public int getNumberOfEnemies() {
+        return numberOfEnemies;
+    }
```
Должен быть (ещё или вместо) статичный метод от `complexity`, который содержит в себе формулу вычисления. 
Сейчас формула у тебя зарыта где-то в коде.

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/game/LevelBuilder.java:
```diff
@@ -65,18 +71,31 @@ public Map build() {
+    // generates door and checks if the player can reach ot from entrance
+    private Position generateDoor(Map map) {
+        Position doorPos = getRandomPosition();
+        while(map.getDistance(doorPos, map.getEntrance()) == -1) {
+            doorPos = getRandomPosition();
+        }
```
Чудовищно, но поправим потом

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/game/unit/Movement.java:
Когда я говорил про Movement, я вообще не это имел ввиду.
Я имел ввиду класс который по структуре точно такой же абсолютно как `Position`, просто применяется там где семантически нужно смещение, а не позиция.

Можно его даже сделать как
```java
    public record Movement(int x, int y) {
        static List<Movement> defaults = Arrays.asList(
                new Movement(-1, 0),
                new Movement(0, -1),
                new Movement(0, +1),
                new Movement(+1, 0)
        );
    }
```
Хотя список значений по умолчанию тут совсем не обязателен

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/state/Map.java:
```diff
@@ -74,4 +91,44 @@ public int hashCode() {
+
+    private record QueueNode(Position pos, int dist) {}
+
+    private boolean isCellValid(Position pos) {
+        return ((pos.getY() >= 0) && (pos.getY() < getHeight())
+                && (pos.getX() >= 0) && (pos.getX() < getWidth()));
+    }
```
Всё ещё странное имя у метода

BurnBirdX7 commented on app/src/main/java/ru/itmo/rogue/model/state/State.java:
```diff
@@ -11,10 +11,9 @@ public class State {
-
-
     public boolean running = true;
     public Judge rdj = new JustJudge();
+    public Map.MapTile doorOut = Map.MapTile.DOOR_IN;
```
Не люблю это поле :{

Ane1y committed:
fixed everything according to the comments of Artemiy Lazarev
Movement now is record and contains shift
No more door out to understand which room ti generate further
And number of enemies is finally in map!!!!

Ane1y committed:
changed ini value for focus in delta
added player position update in delta

Ane1y committed:
deleted some rows

BurnBirdX7 committed:
Refactoring

BurnBirdX7 committed:
Unit tests for map

Ane1y committed:
added one little test

BurnBirdX7 committed:
Unit tests for map
